src/Core/Checkout/Customer/SalesChannel/AccountService.php

```diff
@@ -179,46 +182,94 @@ private function loginByCustomer(CustomerEntity $customer, SalesChannelContext $
 
     /**
      * @throws CustomerNotFoundException
+     * @throws BadCredentialsException
      */
     private function getCustomerByEmail(string $email, SalesChannelContext $context, bool $includeGuest = false): CustomerEntity
     {
         $criteria = new Criteria();
         $criteria->addFilter(new EqualsFilter('email', $email));
-        if (!$includeGuest) {
-            $criteria->addFilter(new EqualsFilter('guest', false));
-        }
-        $criteria->addSorting(new FieldSorting('createdAt'));
-        $criteria->setLimit(1);
-
-        $customer = $this->fetchCustomer($criteria, $context);
 
+        $customer = $this->fetchCustomer($criteria, $context, $includeGuest);
         if ($customer === null) {
             throw new CustomerNotFoundException($email);
         }
 
         return $customer;
     }
 
+    /**
+     * @throws BadCredentialsException
+     */
     private function getCustomerById(string $id, SalesChannelContext $context): CustomerEntity
     {
         $criteria = new Criteria([$id]);
-        $customer = $this->fetchCustomer($criteria, $context);
 
+        $customer = $this->fetchCustomer($criteria, $context, true);
         if ($customer === null) {
             throw new CustomerNotFoundByIdException($id);
         }
 
         return $customer;
     }
 
-    private function fetchCustomer(Criteria $criteria, SalesChannelContext $context): ?CustomerEntity
+    /**
+     * Add only filters to the $criteria for values which have an index in the database, e.g. id, or email,
+     * the rest should be done via PHP. This method filters for the standard customer related constraints
+     * like active or the sales channel assignment.
+     *
+     * @throws BadCredentialsException
+     */
+    private function fetchCustomer(Criteria $criteria, SalesChannelContext $context, bool $includeGuest = false): ?CustomerEntity
     {
-        $criteria->addFilter(new MultiFilter(MultiFilter::CONNECTION_OR, [
-            new EqualsFilter('boundSalesChannelId', null),
-            new EqualsFilter('boundSalesChannelId', $context->getSalesChannel()->getId()),
-        ]));
+        $criteria->setTitle('account-service');
```

Comment:
```suggestion
        $criteria->setTitle('account-service::fetchCustomer');
```

that should make it easier to identify where the query is coming from, `'account-service'` may be a bit too generic